orangecontrib/spectroscopy/widgets/owoverlay.py

```diff
@@ -0,0 +1,281 @@
+                DomainModel.Separator,
+                DomainModel.CLASSES,
+                DomainModel.Separator,
+                DomainModel.METAS,
+            ),
+            valid_types=ContinuousVariable,
+        )
+
+        self.contextAboutToBeOpened.connect(lambda x: self._init_interface_data(x[0]))
+
+        common_options = dict(labelWidth=50, orientation=Qt.Horizontal,
+                              sendSelectedValue=True)
+        
+        box = gui.vBox(self.imagebox)
+
+        self.feature_value = gui.comboBox(
+            box,
+            self,
+            "attr_value",
+            label="Feature:",
+            contentsLength=12,
+            searchable=True,
+            callback=self._update_feature_value,
+            model=self.feature_value_model,
+            **common_options
+        )
+
+        gui.comboBox(
+            box, self, f"attr_x", label=f"Axis X:",
+            callback=self._attr_changed,
+            model=self.xy_model, **common_options
+            )
+        gui.comboBox(
+            box, self, f"attr_y", label=f"Axis Y:",
+            callback=self._attr_changed,
+            model=self.xy_model, **common_options
+            )
+
+        colorsbox = gui.widgetBox(self.controlArea, "Image colors")
+        colorsbox.layout().addWidget(self.imageplot.color_settings_box)
+        
+        gui.rubber(self.controlArea)
+        gui.auto_commit(self.controlArea, self, "autocommit", "Send Data")
+
+        self.imageplot.color_cb.activated.connect(self.commit.deferred)
+        self.imageplot._threshold_low_slider.sliderReleased.connect(self.commit.deferred)
+        self.imageplot._threshold_high_slider.sliderReleased.connect(self.commit.deferred)
+        self.imageplot._level_high_le.editingFinished.connect(self.commit.deferred)
+        self.imageplot._level_low_le.editingFinished.connect(self.commit.deferred)
+        self._init_attr_values(self.data)
+
+    def _attr_changed(self):
+        self._update_attrs()
+        self.commit.deferred()
+
+    def _init_attr_values(self, data):
+        domain = data.domain if data is not None else None
+        self.feature_value_model.set_domain(domain)
+        self.attr_value = (
+            self.feature_value_model[0] if self.feature_value_model else None
+        )
+        self.xy_model.set_domain(domain)
+        self.attr_x = self.xy_model[0] if self.xy_model else None
+        self.attr_y = self.xy_model[1] if len(self.xy_model) >= 2 else self.attr_x
+
+    def _init_interface_data(self, data):
+        self._init_attr_values(data)
+        self.imageplot.init_interface_data(data)
+
+    def _update_feature_value(self):
+        self.redraw_data()
+
+    def _update_attrs(self):
+        self.imageplot.attr_x = self.attr_x
+        self.imageplot.attr_y = self.attr_y
+        self.redraw_data()
+    
+    def redraw_data(self):
+        self.imageplot.update_view()
+        self.commit.deferred()
+    
+    def set_vimage(self, data):
```

Comment:
The reason that tests are failing is that this function is sloppy. This function should be asynchronous (threaded), because it relies on painting that is itself asynchronous. The check `if im is not None` should not be used for avoiding outputs if the image is not yet draw. Image should always exist here unless there was an error.